src/ConstantInterval.cpp

```diff
@@ -584,36 +588,51 @@ ConstantInterval abs(const ConstantInterval &a) {
 }
 
 ConstantInterval operator<<(const ConstantInterval &a, const ConstantInterval &b) {
-    // Try to map this to a multiplication and a division
-    ConstantInterval mul, div;
-    constexpr int64_t one = 1;
-    if (b.min_defined) {
-        if (b.min >= 0 && b.min < 63) {
-            mul.min = one << b.min;
-            mul.min_defined = true;
-            div.max = one;
-            div.max_defined = true;
-        } else if (b.min > -63 && b.min <= 0) {
-            mul.min = one;
-            mul.min_defined = true;
-            div.max = one << (-b.min);
-            div.max_defined = true;
+    // In infinite integers (with no overflow):
+
+    // a << b == a * 2^b
+
+    // This can't be used directly, because if b is negative then 2^b is not an
+    // integer. Instead, we'll break b into a difference of two positive values:
+    // b = b_pos - b_neg
+    // So
+    // a * 2^b
+    // = a * 2^(b_pos - b_neg)
+    // = (a * 2^b_pos) / 2^b_neg
+
+    // From there we can use the * and / operators.
+
+    ConstantInterval b_pos = max(b, 0), b_neg = max(-b, 0);
+
+    // At this point, we have sliced the interval b into two parts. E.g.
+    // if b = [10, 12],  b_pos = [10, 12] and b_neg = [0, 0]
+    // if b = [-4, 8],   b_pos = [0, 8]   and b_neg = [0, 4]
+    // if b = [-10, -3], b_pos = [0, 0]   and b_neg = [3, 10]
+    // if b = [-3, inf], b_pos = [0, inf] and b_neg = [0, 3]
+    // In all cases, note that b_pos - b_neg = b by our definition of - for
+    // ConstantIntervals above.
+
+    auto two_to_the = [](const ConstantInterval &i) {
+        const int64_t one = 1;
+        ConstantInterval r;
+        // We should know i is positive at this point.
+        internal_assert(i.min_defined && i.min >= 0);
+        r.min_defined = true;
+        if (i.min >= 63) {
+            // It's at least a value too large for us to represent, which is not
+            // the same as min_defined = false.
```

Comment:
I'm not sure I understand what this distinction is exactly, and why it's used for `r.min` and not `r.max`

Reply:
!min_defined means either there is no lower bound, or that the lower bound is too negative for us to represent. In either case, we know that, for example, -1000 is in the interval.

In this case there is a lower bound, but it's too *large* for us to represent. So -1000 is definitely not in the interval.

We don't have to worry about the same thing for max_defined, because this code is making large positive values.

Reply:
Ah gotcha